![GPT-Chat Banner](https://openfileserver.chloelavrat.com/workshops/RadioGPT/assets/gpt-chat-banner.png)



> 💡 **PLEASE CONNECT USING A GPU SESSION FOR MORE COMPUTE POWER** :
>
> `Runtime > Change runtime type > T4 GPU > Save`


# 🎯 Objective


The goal of this second part of the workshop is to explore how to interact with and experiment using a pre-trained large language model. Through hands-on experience with a small but still larger than previously pre-trained model (83M parameters). In this section you will discover:

- How to load and work with a pre-trained language model.
- The structure of chatbot datasets and how to leverage them for generating responses.
- The effects of adjusting model parameters like temperature, prompt design, and token limits.
- The challenges of training and using very small models in real-world applications, including biases and overfitting.
- How to explore and critique the model's responses to identify areas for improvement.

By the end of this phase, you will have experimented with generating text, gained insight into model behavior, and understood the trade-offs involved in working with smaller models. 🚀

# 🎬 Notebook initialization

In [4]:
!git clone https://github.com/chloelavrat/RadioGPT.git > /dev/null 2>&1
!cd RadioGPT && git checkout clavrat/first-version > /dev/null 2>&1
!pip install torch datasets tqdm transformers tiktoken> /dev/null 2>&1

# 📀 Dataset Overview

The French Alpaca Dataset contains **110,368 instruction-based examples** in French in the Alpaca format. The dataset was specifically created to fine-tune general language models for tasks like question-answering and text generation.

**Here’s the twist:** this dataset was entirely generated by GPT-3.5-turbo itself! In other words, an LLM was used to create data that would train... another LLM. It's a fascinating example of AI systems helping to improve their own performance, all while producing a diverse set of French instructions that can be used for fine-tuning and model experimentation.

🔗 [Explore the dataset on Hugging Face](https://huggingface.co/datasets/jpacifico/French-Alpaca-dataset-Instruct-110K)

In [2]:
#@title Download French Alpaca dataset
!mkdir dataset
!wget https://openfileserver.chloelavrat.com/workshops/RadioGPT/dataset/Acquiesce_data_110k_instructions.json -O dataset/Acquiesce_data_110k_instructions.json
print("Dataset downloaded")

--2024-12-14 18:13:55--  https://openfileserver.chloelavrat.com/workshops/RadioGPT/dataset/Acquiesce_data_110k_instructions.json
Resolving openfileserver.chloelavrat.com (openfileserver.chloelavrat.com)... 149.202.72.149
Connecting to openfileserver.chloelavrat.com (openfileserver.chloelavrat.com)|149.202.72.149|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 51958855 (50M) [application/json]
Saving to: ‘dataset/Acquiesce_data_110k_instructions.json’

dataset/Acquiesce_d 100%[===================>]  49.55M  8.63MB/s    in 7.2s    

2024-12-14 18:14:03 (6.92 MB/s) - ‘dataset/Acquiesce_data_110k_instructions.json’ saved [51958855/51958855]

Dataset downloaded


In [18]:
#@title Load Alpaca dataset
from RadioGPT.gptmodel.core.dataset import AlpacaDataset

block_size = 64 # you remember, "sets the length of input sequences the model processes at once"

# Dataset
dataset = AlpacaDataset("dataset/Acquiesce_data_110k_instructions.json", block_size)

print("Vocabulary size   : ", dataset.vocab_size)
print("Batch token       : \n", dataset.__getitem__(42))
print("Batch decoded     : \n", dataset.decode(dataset.__getitem__(42)[0].tolist()))

Loaded 453174 blocks
Maximum sequence length: 64
Vocabulary size   :  50257
Batch token       : 
 (tensor([24361,    25,   371, 20954,  2454,   443, 48659,   660,   424,   452,
          415,   551,  2026,   285,  1747,    13,  1004,  4767,   270,  9005,
          390,  3738, 42722,   390,  9281,    12,  3109,   929,  2634,   563,
         1556,   555,   542,    68,  6722,  2350, 45567,  3444, 38599,   300,
            6, 10034, 32177,   288,     6,   403,  4273,   270, 19716, 45567,
          627,  2654,   473,  1410, 14064,   660, 12797, 39349,   288,     6,
         2306,   411,   285]), tensor([   25,   371, 20954,  2454,   443, 48659,   660,   424,   452,   415,
          551,  2026,   285,  1747,    13,  1004,  4767,   270,  9005,   390,
         3738, 42722,   390,  9281,    12,  3109,   929,  2634,   563,  1556,
          555,   542,    68,  6722,  2350, 45567,  3444, 38599,   300,     6,
        10034, 32177,   288,     6,   403,  4273,   270, 19716, 45567,   627,
         265

# 🧠 Model

In [19]:
#@title Loading ...
# Load RadioGPT's checkpoint :)
from RadioGPT.gptmodel.core.model import GPTlite
from RadioGPT.gptmodel.core.utils import load_model
import os, subprocess, torch
# get device
device = (
    torch.device("mps") if torch.backends.mps.is_available() else
    torch.device("cuda") if torch.cuda.is_available() else
    torch.device("cpu")
)

# downloading model
print("Downloading model...")
os.makedirs("models", exist_ok=True)
subprocess.run(["wget", "https://openfileserver.chloelavrat.com/workshops/RadioGPT/models/model_gpt_chat_best.pth", "-O", "models/model_gpt_chat_best.pth"])

def load_model(model_path, device, config):
    # Load the model checkpoint
    checkpoint = torch.load(
        model_path, map_location=device, weights_only=False)

    model = GPTlite(config)
    model.load_state_dict(checkpoint)

    return model

# Hyperparameters
block_size = 64
n_embd = 512
n_head = 8
n_layer = 10
dropout = 0.2

config = {
    'context_size': block_size,
    'vocab_size': dataset.vocab_size,
    'embedding_dim': n_embd,
    'num_heads': n_head,
    'num_layers': n_layer,
    'dropout': dropout
}

# Loading model in memory
print("Loading model...")
model = load_model("models/model_gpt_chat_best.pth", device, config)
model = model.to(device)
print("Model loaded !")

# Get the total number of parameters in the model
total_params = sum(p.numel() for p in model.parameters())

print(f"You will use your {device}")
print(f"Total number of parameters   {total_params / 1e6:.1f}M")

Loading model...
Model loaded !
You will use your cuda
Total number of parameters   83.1M


As you can see below this model version features a lot more parameters resulting in a larger model of **83.1M** Parameters.

Our model has **83.1 million parameters (83.1M)**, which is considered quite small compared to larger GPT models like GPT-2 and GPT-3. Here's a quick comparison:

* **GPT-2 Small:** 117M parameters
* **GPT-2 Medium:** 345M parameters
* **GPT-2 Large:** 762M parameters
* **GPT-3:** 175B parameters

# 🎨 Let's Play

In [20]:
#@title Selected prompt

# play with the model
def generate_response(model, dataset, prompt, device, max_new_tokens):
    model.eval()
    # Encode the prompt
    input_tensor = dataset.encode(prompt).unsqueeze(0).to(device)

    # Generate text using the model's generate method
    with torch.no_grad():
        generated_indices = model.generate(input_tensor, max_new_tokens)
        generated_text = dataset.decode(generated_indices[0].tolist())

    # Return only the newly generated part (after the prompt)
    return generated_text[len(prompt):]


prompt = 'Prépare une recette de pâtes à la carbonara.' # @param ["Prépare une recette de pâtes à la carbonara.", "Quel est l'élément chimique avec le numéro atomique 29 ?", "Rédige un court paragraphe sur le thème de l'amitié et de la confiance."]
max_new_tokens = 106 # @param {type:"slider", min:5, max:500, step:1}


prompt = f"Question: {prompt}\nAnswer:"
prompt = prompt + generate_response(model, dataset, prompt, device, max_new_tokens)
print(prompt)

Question: Prépare une recette de pâtes à la carbonara.
Answer: Ingrédients : - 400 g de spaghetti - 150 g de pancetta - 4 œufs - 150 g de guanciale - 100 g de pecorino râpé - Pelerocke vanille

<|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|>


In [21]:
#@title Let's prompt it!

# play with the model
def generate_response(model, dataset, prompt, device, max_new_tokens):
    model.eval()
    # Encode the prompt
    input_tensor = dataset.encode(prompt).unsqueeze(0).to(device)

    # Generate text using the model's generate method
    with torch.no_grad():
        generated_indices = model.generate(input_tensor, max_new_tokens)
        generated_text = dataset.decode(generated_indices[0].tolist())

    # Return only the newly generated part (after the prompt)
    return generated_text[len(prompt):]


prompt = 'Décris les différences entre le modèle GPT-2 et le modèle GPT-3.' # @param {type:"string"}
max_new_tokens = 180 # @param {type:"slider", min:5, max:500, step:1}


prompt = f"Question: {prompt}\nAnswer:"
prompt = prompt + generate_response(model, dataset, prompt, device, max_new_tokens)
print(prompt)

Question: Décris les différences entre le modèle GPT-2 et le modèle GPT-3.
Answer: Le modèle GPT-2 est un modèle de langage généré développé par OpenAI. Il est basse terme de gestion du temps il ferais parcour 10°Côti du pays de ville extra Texas par le temps.
Answer: Appreneiffelques endives du genre de nombre impair de sportôti endorphives pas endantes du pays prochain nombre ni cultures divers. Par extra pas); scan n extrautes de nombreonde par men ville à 10 heures animives du nombre de sportôtcles par men numvernement nature num Madrid.
Answer: On the humanien nombre de vol extra termes a men par population nombre de nombreux pays participants plus men par le plus haut santes


# ✨ **Conclusion**

This exercise demonstrates both the potential and the limitations of working with smaller models. Despite having only **83.1M parameters**, the model is able to generate coherent sentence structures. However, as seen in the example, its responses can lack meaningful content and are prone to producing nonsensical phrases.

>**Example**  
>**Question:** *Décris les différences entre le modèle GPT-2 et le modèle GPT-3*  
>**Answer:** *Le modèle GPT-2 est un modèle de langage généré développé par OpenAI. Il est basse de pays ose ter voliance par le cad de pays membre longue de bell. Ass end quotid men endse animerie longues de nombreonde extrautes vert endale du musée nume numaraccent.*

While the grammar and structure seem correct, the generated response drifts into gibberish, showcasing the trade-offs between model size and capability. This highlights why larger models are often necessary for more accurate and contextually aware language generation, as they have more capacity to represent complex relationships within the data.

In the next section, a fine-tuned LLM will be used to explore how training on domain-specific data can significantly improve the quality of responses. Let’s take this learning further! 🚀